src/main/scala/diplomacy/LazyModule.scala

```diff
@@ -112,141 +168,257 @@ abstract class LazyModule()(implicit val p: Parameters)
     children.filter(!_.omitGraphML).foreach { c => c.edgesGraphML(buf, pad) }
   }
 
+  /** Execute function by all of this [[LazyModule]]'s  [[children]].
+    *
+    * @param iterfunc function to run
+    * */
   def nodeIterator(iterfunc: (LazyModule) => Unit): Unit = {
     iterfunc(this)
     children.foreach( _.nodeIterator(iterfunc) )
   }
 
+  /** Accessor of val [[children]]. */
   def getChildren = children
 }
 
 object LazyModule
 {
+  /** Current [[LazyModule]] scope. The scope is a stack of [[LazyModule]]/[[LazyScope]].
+    * Each call to [[LazyScope.apply]] or [[LazyModule.apply]] will push that item onto the current scope.
+    * */
   protected[diplomacy] var scope: Option[LazyModule] = None
+  /** Global index of [[LazyModule]], notice there is no zeroth module. */
   private var index = 0
 
+  /** Any instance of [[LazyModule]] should be wrapper in to a [[LazyModule.apply]],
+    * Making use this singleton manage scope and index for elaboration.
+    *
+    * @param bc [[LazyModule]] instance that needs to be wrapped.
+    * @param valName [[ValName]] used to name this instance,
+    *                it can be automatically generated by [[ValName]] macro, or specify manually.
+    * @param sourceInfo [[SourceInfo]] information about where this [[LazyModule]] is being generated
+    * */
   def apply[T <: LazyModule](bc: T)(implicit valName: ValName, sourceInfo: SourceInfo): T = {
-    // Make sure the user put LazyModule around modules in the correct order
-    // If this require fails, probably some grandchild was missing a LazyModule
-    // ... or you applied LazyModule twice
-    require (scope.isDefined, s"LazyModule() applied to ${bc.name} twice ${sourceLine(sourceInfo)}")
+    // Make sure the user puts [[LazyModule]] around modules in the correct order.
+    require (scope.isDefined,
+             s"LazyModule() applied to ${bc.name} twice ${sourceLine(sourceInfo)}. Ensure that descendant LazyModules are instantiated with the LazyModule() wrapper and that you did not call LazyModule() twice.")
     require (scope.get eq bc, s"LazyModule() applied to ${bc.name} before ${scope.get.name} ${sourceLine(sourceInfo)}")
+    // [[LazyModule.scope]] stack pop.
     scope = bc.parent
     bc.info = sourceInfo
     if (!bc.suggestedNameVar.isDefined) bc.suggestName(valName.name)
     bc
   }
 }
 
+/** trait which must be returned in [[LazyModule.module]] for generating actual hardware. */
 sealed trait LazyModuleImpLike extends RawModule
 {
+  /** The [[LazyModule]] which will generate this [[LazyModuleImpLike]].
```

Comment:
You forgot to close this and a few other comments just a bit lower. I noticed this because I was attempting to build the HTML docs with `sbt doc` and it failed to compile.

Reply:
im sure this was my fault. I need to run lint on this

Reply:
I just fixed this.